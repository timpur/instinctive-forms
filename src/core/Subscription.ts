export type Subscriber<TArgs extends any[]> = (...args: TArgs) => void;

export class Subscription<TArgs extends any[]> {
  subscribers: Array<Subscriber<TArgs>> = [];

  subscribe(cb: Subscriber<TArgs>) {
    this.subscribers.push(cb);
    return () => this.unsubscribe(cb);
  }

  unsubscribe(cb: Subscriber<TArgs>) {
    const index = this.subscribers.indexOf(cb);

    if (index !== -1) this.subscribers.splice(index, 1);
  }

  invoke(...args: TArgs) {
    this.subscribers.forEach(subscriber => subscriber(...args));
  }
}

export const connectSubscription = <TArgs extends any[]>(sub1: Subscription<TArgs>, sub2: Subscription<TArgs>) =>
  sub1.subscribe((...args) => sub2.invoke(...args));
