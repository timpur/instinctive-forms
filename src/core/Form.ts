import {
  Callback,
  FormEvents,
  FormValidations,
  TPrimitive,
  FormErrors,
  FormValidationEventTypes,
  FormEvent
} from "../types";
import { runValidation, arrayChanged } from "../helpers";
import { getConfig } from "../index";

export const ERRORS_KEY = "__errors__";

export interface IFormValidationProps {
  disabled?: boolean;
  onChangeValidation?: FormValidations;
  onBlurValidation?: FormValidations;
  onSubmitValidation?: FormValidations;
}

export interface IFormValidation {
  getPath(): string;
  getErrorPath(): string;
  setValidationErrorsIfChanged(eventType: FormValidationEventTypes): void;
  getValidationErrors(eventType: FormValidationEventTypes, value: TPrimitive): FormErrors;
}

export const setValidationErrorsIfChanged = (
  component: IFormValidation,
  props: IFormValidationProps,
  eventType: FormValidationEventTypes
) => {
  const store = getConfig().storeAdapter;
  const path = component.getPath();
  const errorPath = component.getErrorPath();
  const value = store.get(path);
  const currentErrors = store.get(errorPath, []) as FormErrors;
  const newErrors = getValidationErrors(props, eventType, value);

  if (arrayChanged(currentErrors, newErrors)) {
    store.set(errorPath, newErrors);
  }
};

export const getValidationErrors = (
  props: IFormValidationProps,
  eventType: FormValidationEventTypes,
  value: TPrimitive
) => {
  const { disabled, onSubmitValidation, onBlurValidation, onChangeValidation } = props;
  const errors = [];
  if (disabled) return errors;
  switch (eventType) {
    case "runSubmitValidation":
      errors.unshift(...runValidation(onSubmitValidation, value));
    case "runBlurValidation":
      errors.unshift(...runValidation(onBlurValidation, value));
    case "runChangeValidation":
      errors.unshift(...runValidation(onChangeValidation, value));
  }
  return errors;
};

export const filterFormEvents: <TEvent extends { type: TEventType }, TEventType>(
  cb: Callback<[TEvent]>,
  events: TEventType[]
) => (event: FormEvents) => void = (cb, events) => event => {
  if (events.indexOf(event.type as any) !== -1) {
    return cb(event as any);
  }
};

export const validationEvents: FormValidationEventTypes[] = [
  "runChangeValidation",
  "runBlurValidation",
  "runSubmitValidation"
];
export const filterFormValidationEvents: (
  cb: Callback<[FormEvent<{ type: FormValidationEventTypes; sender: any }>]>
) => (event: FormEvents) => void = cb => {
  return filterFormEvents(cb, validationEvents);
};
