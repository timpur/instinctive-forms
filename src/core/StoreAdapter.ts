import { TStoreValue, Callback, Selector } from "../types";
import { ChangeDetection, objectChanged } from "../helpers";

export interface IStoreAdapter<TState = {}> {
  state(): TState;
  get(path: string, defaultValue?: TStoreValue): TStoreValue;
  set(path: string, value: TStoreValue): TStoreValue;
  merge(path: string, value: object | []): TStoreValue;
  setPaths(paths: Array<{ path: string; value: TStoreValue }>): TState;
}

export abstract class StoreAdapter<TState = {}> implements IStoreAdapter<TState> {
  abstract state(): TState;
  abstract subscribe(cb: Callback): Callback;

  abstract _get(path: string, defaultValue?: TStoreValue): TStoreValue;
  abstract _set(path: string, value: TStoreValue);
  abstract _merge(path: string, value: TStoreValue);
  abstract _setPaths(paths: Array<{ path: string; value: TStoreValue }>);

  get(path: string, defaultValue?: TStoreValue) {
    return this._get(path, defaultValue);
  }
  set(path: string, value: TStoreValue) {
    this._set(path, value);
    return this.get(path);
  }
  merge(path: string, value: object | []) {
    this._merge(path, value);
    return this.get(path);
  }
  setPaths(paths: Array<{ path: string; value: TStoreValue }>) {
    this._setPaths(paths);
    return this.state();
  }
}

export const StoreConnect = (storeAdapter: StoreAdapter, update: Callback, ...paths: string[]) => {
  const getCurrentState: Selector<object> = () =>
    paths
      .map(path => [path, storeAdapter.get(path)])
      .reduce((state, [path, value]) => {
        state[path as string] = value;
        return state;
      }, {});
  const changeDetection = new ChangeDetection(getCurrentState(), objectChanged);
  return storeAdapter.subscribe(() => {
    if (changeDetection.check(getCurrentState())) update();
  });
};
