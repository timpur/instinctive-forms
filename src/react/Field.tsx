import { createElement, Component, Fragment, ReactNode } from "react";
import { FormContext, IFormContext } from "./Form";
import { Callback, TPrimitive, FormValidationType, FormValidations, FormErrors, FormFilters } from "../types";
import { getConfig } from "../index";
import { StoreConnect } from "../core/StoreAdapter";
import {
  buildPath,
  runValidation,
  runFilters,
  arrayChanged,
  RenderChildren,
  renderChildrenWithProps,
  validateNameProp
} from "../helpers";
import { ERRORS_KEY } from "../constants";

export interface IFieldChildProps {
  path: string;
  errorPath: string;
  value: TPrimitive;
  errors: FormErrors;
  onChange: Callback<[TPrimitive]>;
  onBlur: Callback;
}

export interface IFieldProps {
  name: string;
  path?: string;
  onChangeValidation?: FormValidations;
  onBlurValidation?: FormValidations;
  onSubmitValidation?: FormValidations;
  filters?: FormFilters;
  disabled?: boolean;
  children?: RenderChildren<IFieldChildProps, ReactNode>;
}

export class Field extends Component<IFieldProps, {}> {
  private unsubscribeFromStore: Callback;
  private unsubscribeFromForm: Callback;

  constructor(...args: [any, any]) {
    super(...args);

    if (!this.context.type) throw new Error("Field must be rendered within a Form.");
    validateNameProp(this.props.name);
    this.connectToStore();
    const { subscription } = this.context as IFormContext;
    this.unsubscribeFromForm = subscription.subscribe(event => this.setValidationErrorsIfChanged(event));
  }

  componentWillReceiveProps(nextProps: IFieldProps, nextContext: IFormContext) {
    if (this.getPath() !== this.getPath(nextProps, nextContext)) {
      this.connectToStore(nextProps, nextContext);
    }
    if (this.props.disabled !== nextProps.disabled) {
      this.setValidationErrorsIfChanged();
    }
  }

  componentWillUnmount = () => {
    this.unsubscribeFromStore();
    this.unsubscribeFromForm();
  };

  connectToStore = (props: IFieldProps = this.props, context: IFormContext = this.context) => {
    if (this.unsubscribeFromStore) this.unsubscribeFromStore();
    const store = getConfig().storeAdapter;
    this.unsubscribeFromStore = StoreConnect(
      store,
      () => this.forceUpdate(),
      this.getPath(props, context),
      this.getErrorPath(props, context)
    );
  };

  getPath = (props: IFieldProps = this.props, context: IFormContext = this.context) => {
    const { path, name } = props;
    const { path: contextPath } = context;
    return buildPath(path || contextPath, name);
  };

  getErrorPath = (props: IFieldProps = this.props, context: IFormContext = this.context) => {
    const { path, name } = props;
    const { path: contextPath } = context;
    return buildPath(path || contextPath, ERRORS_KEY, name);
  };

  onChange = (_value: TPrimitive) => {
    const value = runFilters(this.props.filters, _value);
    const errors = this.getValidationErrors("onChange", value);

    const store = getConfig().storeAdapter;
    store.setPaths([{ path: this.getPath(), value: value }, { path: this.getErrorPath(), value: errors }]);
  };

  onBlur = () => {
    this.setValidationErrorsIfChanged("onBlur");
  };

  setValidationErrorsIfChanged = (type: FormValidationType = "onChange") => {
    const store = getConfig().storeAdapter;
    const value = store.get(this.getPath());
    const currentErrors = store.get(this.getErrorPath()) as FormErrors;
    const newErrors = this.getValidationErrors(type, value);

    if (arrayChanged(currentErrors, newErrors)) {
      store.set(this.getErrorPath(), newErrors);
    }
  };

  getValidationErrors = (type: FormValidationType, value: TPrimitive) => {
    const { disabled, onSubmitValidation, onBlurValidation, onChangeValidation } = this.props;
    const errors: FormErrors = [];
    if (disabled) return errors;

    switch (type) {
      case "onSubmit":
        errors.unshift(...runValidation(onSubmitValidation, value));
      case "onBlur":
        errors.unshift(...runValidation(onBlurValidation, value));
      case "onChange":
        errors.unshift(...runValidation(onChangeValidation, value));
    }
    return errors;
  };

  render() {
    const store = getConfig().storeAdapter;
    const path = this.getPath();
    const errorPath = this.getErrorPath();
    const value = store.get(path);
    const errors = store.get(errorPath, []) as FormErrors;
    return (
      <Fragment>
        {renderChildrenWithProps<IFieldChildProps>(this.props.children, {
          path,
          errorPath,
          value,
          errors,
          onChange: this.onChange,
          onBlur: this.onBlur
        })}
      </Fragment>
    );
  }

  static contextType = FormContext;
}
