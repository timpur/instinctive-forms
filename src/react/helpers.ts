import { ReactElement, ComponentType } from "react";
import { Omit } from "../types";

export type Child<P = any> = ((props: P) => any) | ReactElement<any>;

export const renderChildrenWithProps = <P = {}>(children: Child<P> | Array<Child<P>>, props: P) => {
  const mapChild = child => {
    if (typeof child === "function") return child(props);
    return child;
  };

  return Array.isArray(children) ? children.map(mapChild) : mapChild(children);
};

export type HOC<N> = <P extends N, R = Omit<P, keyof N>>(Component: ComponentType<P>) => ComponentType<R>;
