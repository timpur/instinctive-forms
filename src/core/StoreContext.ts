import { IStoreAdapter, StoreAdapter } from "./StoreAdapter";
import { buildPath } from "../helpers";
import { TStoreValue } from "../types";
import { getConfig } from "../index";

export class StoreContext<TState = {}, TStoreState = {}> implements IStoreAdapter<TState> {
  public readonly adapter: StoreAdapter<TStoreState>;
  public readonly basePath: string;

  constructor(path: string, adapter?: StoreAdapter<TStoreState>) {
    this.basePath = path;
    this.adapter = adapter || getConfig().storeAdapter;
  }

  state(): TState {
    return this.adapter.get(this.basePath) as any;
  }

  get(path: string, defaultValue?: TStoreValue) {
    const fullPath = buildPath(this.basePath, path);
    return this.adapter.get(fullPath, defaultValue);
  }
  set(path: string, value: TStoreValue) {
    const fullPath = buildPath(this.basePath, path);
    return this.adapter.set(fullPath, value);
  }
  merge(path: string, value: object | []) {
    const fullPath = buildPath(this.basePath, path);
    return this.adapter.merge(fullPath, value);
  }
  setPaths(paths: Array<{ path: string; value: TStoreValue }>) {
    const fullPaths = paths.map(({ path, value }) => ({ path: buildPath(this.basePath, path), value }));
    this.adapter.setPaths(fullPaths);
    return this.state();
  }
}
