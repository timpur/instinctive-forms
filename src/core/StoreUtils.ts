import * as dotProp from "dot-prop-immutable";
import { TStoreValue } from "../types";
import { isObject, isEmptyValue } from "../helpers";

const getParentPath = (path: string) =>
  path
    .split(".")
    .slice(0, -1)
    .join(".");
const isRoot = (path: string = "") => path.length === 0;

export const cleanUpEmptyParentsRecursively = (state: any, path: string) => {
  const value = get(state, path);
  const needToRemoveProp = isEmptyValue(value) && value !== null;
  // no clean up needed, since value has a valid value
  if (!needToRemoveProp) return state;
  const cleanedState = dotProp.delete(state, path); // https://github.com/debitoor/dot-prop-immutable/pull/28
  // if path is at root then clean up is done
  if (isRoot(path)) return cleanedState;
  // since were not at the root, keep cleaning up parents
  const parentPath = getParentPath(path);
  return cleanUpEmptyParentsRecursively(cleanedState, parentPath);
};

const get = (state: TStoreValue, path: string, defaultValue?: TStoreValue): TStoreValue => {
  if (!path) return state;
  return dotProp.get(state, path, defaultValue);
};

const set = <TState = TStoreValue>(state: TState, path: string, value: TStoreValue): TState => {
  if (!path) throw new Error("Path can not be undefined, null or empty");
  const changedState = dotProp.set(state, path, value);
  return cleanUpEmptyParentsRecursively(changedState, path);
};

const merge = <TState = TStoreValue>(state: TState, path: string, value: object | []): TState => {
  let changedState;
  if (!path) changedState = dotProp.merge({ state }, "state", value).state;
  else changedState = dotProp.merge(state, path, value);
  return cleanUpEmptyParentsRecursively(changedState, path);
};

const setPaths = <TState = TStoreValue>(state: TState, paths: Array<{ path: string; value: TStoreValue }>): TState => {
  return paths.reduce((obj, { path, value }) => StoreUtils.set(obj, path, value), state);
};

const findPropFormChildren = (
  state: TStoreValue,
  prop: string,
  path: string = null
): Array<{ path: string; value: TStoreValue }> => {
  if (!isObject(state)) return [];

  const keys = Object.keys(state);
  return keys
    .map(stateKey => {
      const newPath = path ? `${path}.${stateKey}` : stateKey;
      if (stateKey === prop) return [{ path: newPath, value: state[stateKey] }];
      else if (isObject(state[stateKey])) {
        return findPropFormChildren(state[stateKey], prop, newPath);
      }
    })
    .reduce((result, items) => (items ? [...result, ...items] : result), []);
};

export const StoreUtils = {
  get,
  set,
  merge,
  setPaths,
  findPropFormChildren
};
