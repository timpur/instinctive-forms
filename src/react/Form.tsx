import { createElement, Fragment, Component, createContext, Context, ReactNode, ComponentType } from "react";
import { Callback, FormSubscription, TStoreValue, Omit } from "../types";
import { StoreContext } from "../core/StoreContext";
import { buildPath, renderChildrenWithProps, RenderChildren, validateNameProp } from "../helpers";
import { getConfig } from "../index";
import { Subscription, connectSubscription } from "../core/Subscription";
import { StoreConnect } from "../core/StoreAdapter";
import { StoreUtils } from "../core/StoreUtils";
import { ERRORS_KEY } from "../constants";

export type HOC<N> = <P extends N, R = Omit<P, keyof N>>(Component: ComponentType<P>) => ComponentType<R>;

export interface IFormContext {
  type: string;
  parent: IFormContext;
  path: string;
  subscription: FormSubscription;
}

export const FormContext: Context<IFormContext> = createContext({
  type: null,
  parent: null,
  path: null,
  subscription: null
});

export interface IFormProps {
  name: string;
  path?: string;
  onSubmit?: Callback<[React.FormEvent<HTMLFormElement>]>;
}

export class Form extends Component<IFormProps, {}> {
  private subscription: FormSubscription;

  constructor(...args: [any, any]) {
    super(...args);

    if (this.context.type) throw new Error("Form can not be place within a form, use Fieldset.");
    validateNameProp(this.props.name);
    this.subscription = new Subscription();
  }

  onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    const { onSubmit } = this.props;
    this.subscription.invoke("onSubmit");
    if (onSubmit) onSubmit(event);
  };

  render() {
    const { path, name } = this.props;
    return (
      <FormContext.Provider
        value={{
          type: "form",
          parent: null,
          path: buildPath(path, name),
          subscription: this.subscription
        }}
      >
        <form name={name} onSubmit={this.onSubmit}>
          {this.props.children}
        </form>
      </FormContext.Provider>
    );
  }

  static contextType = FormContext;
}

export interface IFieldsetProps {
  name: string;
  path?: string;
}

export class Fieldset extends Component<IFieldsetProps, {}> {
  private subscription: FormSubscription;
  private unsubscribeFromForm: Callback;

  constructor(...args: [any, any]) {
    super(...args);

    if (!this.context.type) throw new Error("Fieldset must be rendered within a Form.");
    validateNameProp(this.props.name);
    this.subscription = new Subscription();
    const { subscription } = this.context as IFormContext;
    this.unsubscribeFromForm = connectSubscription(subscription, this.subscription);
  }

  componentWillUnmount() {
    this.unsubscribeFromForm();
  }

  getPath = () => {
    const { path, name } = this.props;
    const { path: contextPath } = this.context as IFormContext;
    return buildPath(path || contextPath, name);
  };

  render() {
    const path = this.getPath();
    const { name } = this.props;
    return (
      <FormContext.Provider
        value={{
          type: "fieldset",
          parent: this.context,
          path,
          subscription: this.subscription
        }}
      >
        <fieldset name={name}>{this.props.children}</fieldset>
      </FormContext.Provider>
    );
  }

  static contextType = FormContext;
}

export interface IFormChildProps extends IFormContext {
  context: StoreContext;
  contextState: TStoreValue;
  contextInError: boolean;
}

export interface IFormConsumerProps {
  children: RenderChildren<IFormChildProps, ReactNode>;
}
export class FormConsumer extends Component<IFormConsumerProps, {}> {
  private unsubscribeFromStore: Callback;

  constructor(...args: [any, any]) {
    super(...args);

    if (!this.context.type) throw new Error("FormConsumer must be rendered within a Form.");
    const { path } = this.context as IFormContext;
    this.connectToStore(path);
  }

  componentWillUnmount() {
    this.unsubscribeFromStore();
  }

  componentWillReceiveProps(nextProps: IFormConsumerProps, nextContext: IFormContext) {
    if (this.context.path !== nextContext.path) {
      const { path } = nextContext;
      this.connectToStore(path);
    }
  }

  connectToStore = path => {
    if (this.unsubscribeFromStore) this.unsubscribeFromStore();
    const store = getConfig().storeAdapter;
    this.unsubscribeFromStore = StoreConnect(store, () => this.forceUpdate(), path);
  };

  render() {
    const { type, parent, path, subscription } = this.context as IFormContext;
    const context = new StoreContext(path);
    const contextState = context.state();
    const contextInError = StoreUtils.findPropFormChildren(contextState, ERRORS_KEY).some(item => !!item.value);
    return (
      <Fragment>
        {renderChildrenWithProps<IFormChildProps>(this.props.children, {
          type,
          parent,
          path,
          subscription,
          context,
          contextState,
          contextInError
        })}
      </Fragment>
    );
  }

  static HOC: HOC<IFormChildProps> = Component => props => (
    <FormConsumer>{formProps => <Component {...formProps} {...props as any} />}</FormConsumer>
  );
  static contextType = FormContext;
}
