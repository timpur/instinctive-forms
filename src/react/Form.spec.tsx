import { createElement, FunctionComponent } from "react";
import { create } from "react-test-renderer";
import { Form, FormContext, Fieldset, FormConsumer, IFormChildProps } from "./Form";
import { setup } from "../index";
import { TestAdapter } from "../../tests/TestAdapter";
import { Subscription } from "../core/Subscription";
import { StoreAdapter } from "../core/StoreAdapter";

const getMockFunctionParam = (fn, call = 1, param = 1) => fn.mock.calls[call - 1][param - 1];

describe("Form", () => {
  it("should render from and children", () => {
    const component = create(
      <Form name="form">
        <p>test</p>
      </Form>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  it("should not allow form inside a form", () => {
    const render = () =>
      create(
        <Form name="form">
          <Form name="form2">
            <p>test</p>
          </Form>
        </Form>
      );

    expect(render).toThrowError("Form can not be place within a form, use Fieldset.");
  });
  it("should not allow path as a name", () => {
    const createComponent = () =>
      create(
        <Form name="path.form">
          <div>test</div>
        </Form>
      );

    expect(createComponent).toThrowError("Name must not be a path");
  });
  it("should provide form context", () => {
    const cb = jest.fn();
    create(
      <Form name="form">
        <FormContext.Consumer>{props => cb(props)}</FormContext.Consumer>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({ type: "form", path: "form", parent: null });
  });
  it("should test path prop", () => {
    const cb = jest.fn();
    create(
      <Form path="path" name="form">
        <FormContext.Consumer>{props => cb(props)}</FormContext.Consumer>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({ type: "form", path: "path.form" });
  });
  it("should submit form", () => {
    const onSubmit = jest.fn();
    const component = create(
      <Form name="form" onSubmit={() => onSubmit("userCallback")}>
        <FormContext.Consumer>
          {props => {
            props.subscription.subscribe(onSubmit);
            return null;
          }}
        </FormContext.Consumer>
      </Form>
    );
    const tree = component.toJSON();

    tree.props.onSubmit();
    expect(onSubmit).toHaveBeenCalledWith("onSubmit");
    expect(onSubmit).toHaveBeenCalledWith("userCallback");
  });
});

describe("FieldSet", () => {
  it("should not allow fieldset outside of a form", () => {
    const render = () => create(<Fieldset name="fieldset" />);

    expect(render).toThrowError("Fieldset must be rendered within a Form.");
  });
  it("should not allow path as a name", () => {
    const createComponent = () =>
      create(
        <Form name="form">
          <Fieldset name="path.fieldset">
            <div>test</div>
          </Fieldset>
        </Form>
      );

    expect(createComponent).toThrowError("Name must not be a path");
  });
  it("should render from, fieldset and children", () => {
    const component = create(
      <Form name="form">
        <Fieldset name="fieldset">
          <p>test</p>
        </Fieldset>
      </Form>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  it("should provide fieldset context", () => {
    const cb = jest.fn();
    create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormContext.Consumer>{props => cb(props)}</FormContext.Consumer>
        </Fieldset>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      type: "fieldset",
      path: "form.fieldset",
      parent: { type: "form", path: "form", parent: null }
    });
  });
  it("should test path prop", () => {
    const cb = jest.fn();
    create(
      <Form path="path" name="form">
        <Fieldset name="fieldset">
          <FormContext.Consumer>{props => cb(props)}</FormContext.Consumer>
        </Fieldset>
        <Fieldset path="path" name="fieldset">
          <FormContext.Consumer>{props => cb(props)}</FormContext.Consumer>
        </Fieldset>
      </Form>
    );

    expect(cb).toBeCalledTimes(2);
    expect(getMockFunctionParam(cb)).toMatchObject({ type: "fieldset", path: "path.form.fieldset" });
    expect(getMockFunctionParam(cb, 2)).toMatchObject({ type: "fieldset", path: "path.fieldset" });
  });
  it("should submit fieldset", () => {
    const onSubmit = jest.fn();
    const component = create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormContext.Consumer>
            {props => {
              props.subscription.subscribe(onSubmit);
              return null;
            }}
          </FormContext.Consumer>
        </Fieldset>
      </Form>
    );
    const tree = component.toJSON();

    tree.props.onSubmit();
    expect(onSubmit).toHaveBeenCalledWith("onSubmit");
  });

  it("should unsubscribe from form on unmount", () => {
    const cb = jest.fn();
    const component = create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormContext.Consumer>{props => cb(props)}</FormContext.Consumer>
        </Fieldset>
      </Form>
    );
    const sub: Subscription<any> = getMockFunctionParam(cb).parent.subscription;
    expect(sub).toBeInstanceOf(Subscription);
    expect(sub.subscribers).toHaveLength(1);
    component.unmount();
    expect(sub.subscribers).toHaveLength(0);
  });
});

describe("Form Consumer", () => {
  let store: StoreAdapter = null;
  beforeEach(() => {
    const testAdapter = new TestAdapter();
    setup(testAdapter);
    testAdapter.set("form.fieldset.test", "test");
    testAdapter.set("path.form.fieldset.test", "path -> test");
    store = testAdapter;
  });

  it("should not allow form consumer outside of a form", () => {
    const render = () => create(<FormConsumer>{() => null}</FormConsumer>);

    expect(render).toThrowError("FormConsumer must be rendered within a Form.");
  });
  it("should render form, fieldset, consumer and child", () => {
    const component = create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormConsumer>{() => <p>test</p>}</FormConsumer>
        </Fieldset>
      </Form>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  it("should have progressive path", () => {
    const cb = jest.fn();
    create(
      <Form name="form">
        <Fieldset name="fieldset1">
          <Fieldset name="fieldset2">
            <FormConsumer>{props => cb(props)}</FormConsumer>
          </Fieldset>
        </Fieldset>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      type: "fieldset",
      path: "form.fieldset1.fieldset2",
      parent: {
        type: "fieldset",
        path: "form.fieldset1",
        parent: { type: "form", path: "form", parent: null }
      }
    });
  });
  it("should have context state", () => {
    const cb = jest.fn();
    create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormConsumer>{props => cb(props)}</FormConsumer>
        </Fieldset>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      contextState: { test: "test" }
    });
  });
  it("should set context state", () => {
    const cb = jest.fn();
    create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormConsumer>{props => cb(props)}</FormConsumer>
        </Fieldset>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      contextState: { test: "test" }
    });

    store.set("form.fieldset.test", "changed");

    expect(cb).toBeCalledTimes(2);
    expect(getMockFunctionParam(cb, 2)).toMatchObject({
      contextState: { test: "changed" }
    });
  });
  it("should deal with changing paths", () => {
    const Test: FunctionComponent<{ path?: string }> = props => (
      <Form path={props.path} name="form">
        <Fieldset name="fieldset">
          <FormConsumer>{props => cb(props)}</FormConsumer>
        </Fieldset>
      </Form>
    );
    const cb = jest.fn();
    const component = create(<Test />);

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      path: "form.fieldset",
      contextState: { test: "test" }
    });

    component.update(<Test />);

    expect(cb).toBeCalledTimes(2);
    expect(getMockFunctionParam(cb, 2)).toMatchObject({
      path: "form.fieldset",
      contextState: { test: "test" }
    });

    component.update(<Test path="path" />);

    expect(cb).toBeCalledTimes(3);
    expect(getMockFunctionParam(cb, 3)).toMatchObject({
      path: "path.form.fieldset",
      contextState: { test: "path -> test" }
    });

    store.set("path.form.fieldset.test", "path -> changed");

    expect(cb).toBeCalledTimes(4);
    expect(getMockFunctionParam(cb, 4)).toMatchObject({
      path: "path.form.fieldset",
      contextState: { test: "path -> changed" }
    });
  });
  it("should not render more times than expected", () => {
    const cb = jest.fn();
    create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormConsumer>{props => cb(props)}</FormConsumer>
        </Fieldset>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      contextState: { test: "test" }
    });

    store.set("form.fieldset.test", "changed");

    expect(cb).toBeCalledTimes(2);
    expect(getMockFunctionParam(cb, 2)).toMatchObject({
      contextState: { test: "changed" }
    });

    store.set("path.form.fieldset.test", "changed");

    expect(cb).toBeCalledTimes(2);
  });

  it("should unsubscribe from store on unmount", () => {
    const cb = jest.fn();
    const component = create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormConsumer>{props => cb(props)}</FormConsumer>
        </Fieldset>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      contextState: { test: "test" }
    });

    store.set("form.fieldset.test", "changed");

    expect(cb).toBeCalledTimes(2);
    expect(getMockFunctionParam(cb, 2)).toMatchObject({
      contextState: { test: "changed" }
    });

    component.unmount();

    store.set("form.fieldset.test", "not seen");
    expect(cb).toBeCalledTimes(2);
  });
  it("should have form in error", () => {
    const cb = jest.fn();
    create(
      <Form name="form">
        <Fieldset name="fieldset">
          <FormConsumer>{props => cb(props)}</FormConsumer>
        </Fieldset>
      </Form>
    );

    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      contextInError: false
    });

    store.set("form.fieldset.__errors__.test", ["error"]);

    expect(cb).toBeCalledTimes(2);
    expect(getMockFunctionParam(cb, 2)).toMatchObject({
      contextInError: true
    });
  });
  it("should pass context via HOC", () => {
    const cb = jest.fn();
    const Test: FunctionComponent<IFormChildProps & { testProp: string }> = props => {
      cb(props);
      return <div>test component: {props.testProp}</div>;
    };
    const ConnectedTest = FormConsumer.HOC(Test);
    const component = create(
      <Form name="form">
        <Fieldset name="fieldset">
          <ConnectedTest testProp="bla" />
        </Fieldset>
      </Form>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
    expect(cb).toBeCalledTimes(1);
    expect(getMockFunctionParam(cb)).toMatchObject({
      path: "form.fieldset",
      contextState: { test: "test" }
    });
  });
});
