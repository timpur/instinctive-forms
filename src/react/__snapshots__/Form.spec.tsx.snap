// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldSet should render from, fieldset and children 1`] = `
<form
  name="form"
  onSubmit={[Function]}
>
  <fieldset
    name="fieldset"
  >
    <p>
      test
    </p>
  </fieldset>
</form>
`;

exports[`Form Consumer should pass context via HOC 1`] = `
<form
  name="form"
  onSubmit={[Function]}
>
  <fieldset
    name="fieldset"
  >
    <div>
      test component: 
      bla
    </div>
  </fieldset>
</form>
`;

exports[`Form Consumer should render form, fieldset, consumer and child 1`] = `
<form
  name="form"
  onSubmit={[Function]}
>
  <fieldset
    name="fieldset"
  >
    <p>
      test
    </p>
  </fieldset>
</form>
`;

exports[`Form should render from and children 1`] = `
<form
  name="form"
  onSubmit={[Function]}
>
  <p>
    test
  </p>
</form>
`;
